From 35872a3956125e8f9569789c822cb03c16001e08 Mon Sep 17 00:00:00 2001
From: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Date: Tue, 19 Sep 2023 13:03:33 +0800
Subject: [PATCH 24/28] configs: arm64: add audio cell config for imx93

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
(configs: arm64: imx93: add SAI{1,2,3} IRQs/mmu, WAKEUPMIX1 mmu and ANA_PLL to the audio application)
Signed-off-by: Zineb El Azzouzi <zineb.elazzouzi@nxp.com>
---
 configs/arm64/imx93-harpoon-config-audio.h   | 52 ++++++++++++++++++++
 configs/arm64/imx93-harpoon-freertos-audio.c | 23 +++++++++
 configs/arm64/imx93-harpoon-rtos.h           | 27 +++++++---
 configs/arm64/imx93-harpoon-zephyr-audio.c   | 23 +++++++++
 4 files changed, 118 insertions(+), 7 deletions(-)
 create mode 100644 configs/arm64/imx93-harpoon-config-audio.h
 create mode 100644 configs/arm64/imx93-harpoon-freertos-audio.c
 create mode 100644 configs/arm64/imx93-harpoon-zephyr-audio.c

diff --git a/configs/arm64/imx93-harpoon-config-audio.h b/configs/arm64/imx93-harpoon-config-audio.h
new file mode 100644
index 00000000..ea7f2a8d
--- /dev/null
+++ b/configs/arm64/imx93-harpoon-config-audio.h
@@ -0,0 +1,52 @@
+/*
+ * i.MX93 family targets - Audio config
+ *
+ * Copyright 2023 NXP
+ *
+ * Authors:
+ *  Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
+ *
+ * This work is licensed under the terms of the GNU GPL, version 2.  See
+ * the COPYING file in the top-level directory.
+ */
+
+#define CONFIG_INMATE_BASE		RTOS_INMATE_BASE
+
+/* Memory & peripherals */
+#define CONFIG_INMATE_REGIONS                   \
+	RTOS_COMMON_MEMORY_REGIONS,                               \
+	RTOS_MEMRAM_MEMORY_REGIONS,                               \
+	RTOS_AUDIO_MEMORY_REGIONS
+
+/* GIC */
+#define CONFIG_INMATE_IRQCHIPS_ADDR             \
+	CONFIG_RTOS_IRQCHIPS_ADDR
+#define CONFIG_INMATE_IRQCHIPS                  \
+	IRQCHIP0(CONFIG_INMATE_IRQCHIPS_ADDR,                     \
+		/* interrupts 32..63 */                                  \
+		RTOS_COMMON_IRQCHIP0_BITMAP1 |                           \
+		RTOS_AUDIO_IRQCHIP0_BITMAP1,                             \
+		/* interrupts 64..95 */                                  \
+		RTOS_COMMON_IRQCHIP0_BITMAP2 |                           \
+		RTOS_AUDIO_IRQCHIP0_BITMAP2,                             \
+		/* interrupts 96..127 */                                 \
+		RTOS_COMMON_IRQCHIP0_BITMAP3 |                           \
+		RTOS_AUDIO_IRQCHIP0_BITMAP3,                             \
+		/* interrupts 128..159 */                                \
+		RTOS_COMMON_IRQCHIP0_BITMAP4 |                           \
+		RTOS_AUDIO_IRQCHIP0_BITMAP4),                            \
+	IRQCHIP1(CONFIG_INMATE_IRQCHIPS_ADDR,                     \
+		/* interrupts 160..191 */                                \
+		RTOS_COMMON_IRQCHIP1_BITMAP1,                            \
+		/* interrupts 192..223 */                                \
+		RTOS_COMMON_IRQCHIP1_BITMAP2 |                           \
+		RTOS_AUDIO_IRQCHIP1_BITMAP2,                             \
+		/* interrupts 224..255 */                                \
+		RTOS_COMMON_IRQCHIP1_BITMAP3,                            \
+		/* interrupts 256..287 */                                \
+		RTOS_COMMON_IRQCHIP1_BITMAP4)
+
+/* IVSHMEM PCI */
+#define CONFIG_INMATE_PCI_DEVICES                    \
+	RTOS_COMMON_PCI_IVSHMEM
+
diff --git a/configs/arm64/imx93-harpoon-freertos-audio.c b/configs/arm64/imx93-harpoon-freertos-audio.c
new file mode 100644
index 00000000..cdbc566f
--- /dev/null
+++ b/configs/arm64/imx93-harpoon-freertos-audio.c
@@ -0,0 +1,23 @@
+/*
+ * i.MX93 target - FreeRTOS audio config
+ *
+ * Copyright 2023 NXP
+ *
+ * Authors:
+ *  Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
+ *
+ * This work is licensed under the terms of the GNU GPL, version 2.  See
+ * the COPYING file in the top-level directory.
+ */
+
+#include "cell-helper.h"
+#include "imx93-harpoon-rtos.h"
+
+/* Name, cores */
+#define CONFIG_INMATE_NAME		"freertos"
+#define CONFIG_INMATE_CORE_BITMAP	(0b10)
+
+/* Base, memory & peripherals, GIC, IVSHMEM PCI */
+#include "imx93-harpoon-config-audio.h"
+
+#include "cell-create.h"
diff --git a/configs/arm64/imx93-harpoon-rtos.h b/configs/arm64/imx93-harpoon-rtos.h
index ddece671..6bfefcf9 100644
--- a/configs/arm64/imx93-harpoon-rtos.h
+++ b/configs/arm64/imx93-harpoon-rtos.h
@@ -12,14 +12,18 @@
 
 typedef enum IRQn {
   /* Device specific interrupts */
+  LPI2C1_IRQn                  = 45,               /**< Low Power Inter-Integrated Circuit module 1 */
   LPUART2_IRQn                 = 52,               /**< Low Power UART 2 */
   TPM2_IRQn                    = 69,               /**< Timer PWM Module (TPM) 2 */
+  SAI1_IRQn                    = 77,              /**< SAI1 45 Receive / Transmit Interrupt */
   CAN_FD2_IRQn                 = 83,               /**< CAN2 interrupt */
   CAN_FD2_ERROR_IRQn           = 84,               /**< CAN2 error interrupt */
   TPM4_IRQn                    = 108,              /**< Timer PWM Module (TPM) 4 */
   SW_MB_L_IRQn                 = 200,              /**< Unused SPI 168, to be used by generic software mailbox in Linux */
   SW_MB_R_IRQn                 = 201,              /**< Unused SPI 169, to be used by generic software mailbox in RTOS */
   ENET_IRQn                    = 213,              /**< ENET1 MAC Interrupt */
+  SAI2_IRQn                    = 202,              /**< SAI2 170 Receive / Transmit Interrupt */
+  SAI3_IRQn                    = 203,              /**< SAI3 171 Receive / Transmit Interrupt */
   ENET_QOS_PMT_IRQn            = 215,              /**< ENET QOS TSN Interrupt from PMT */
   ENET_QOS_IRQn                = 216,              /**< ENET QOS TSN LPI RX exit/Host System/RX/TX Channels[4:0] Interrupt */
 } IRQn_Type;
@@ -49,13 +53,21 @@ typedef enum IRQn {
 
 #define RTOS_AVB_MEMORY_REGIONS \
 	RTOS_AUDIO_MEMORY_REGIONS,                                     \
-	MMIO_REGION_RW( 0x428a0000, 0x428a0000, KB(64)) /* ENET_Q*/    \
-	MMIO_REGION_RWS( 0x42420000, 0x42420000, KB(64)) /* WAKEUPIMX1*/
+	MMIO_REGION_RW( 0x428a0000, 0x428a0000, KB(64)) /* ENET_Q*/
 
 #define RTOS_INDUS_MEMORY_REGIONS \
 	MMIO_REGION_RW( 0x425b0000, 0x425b0000, KB(64)), /* FLEXCAN2*/  \
-	MMIO_REGION_RW( 0x428a0000, 0x428a0000, KB(64)), /* ENET_Q*/    \
-	MMIO_REGION_RWS( 0x42420000, 0x42420000, KB(64)) /* WAKEUPIMX1*/
+	MMIO_REGION_RW( 0x428a0000, 0x428a0000, KB(64)), /* ENET_Q*/   \
+	MMIO_REGION_RWS(0x42420000, 0x42420000, KB(64)), /* WAKEUPMIX1*/\
+	MMIO_REGION_RWS(0x44480000, 0x44480000, KB(8))   /* ANA_PLL */
+
+#define RTOS_AUDIO_MEMORY_REGIONS \
+	MMIO_REGION_RWS(0x42420000, 0x42420000, KB(64)),  /* WAKEUPMIX1*/\
+	MMIO_REGION_RW( 0x42650000, 0x42650000, KB(64)),  /* SAI2 */   \
+	MMIO_REGION_RW( 0x42660000, 0x42660000, KB(64)),  /* SAI3 */   \
+	MMIO_REGION_RW( 0x44340000, 0x44340000, KB(64)),  /* LPI2C1 */ \
+	MMIO_REGION_RW( 0x443b0000, 0x443b0000, KB(64)),  /* SAI1 */   \
+	MMIO_REGION_RWS(0x44480000, 0x44480000, KB(8))    /* ANA_PLL */
 
 #define RTOS_VIRTIO_MEMORY_REGIONS \
 	MMIO_REGION_RW( 0x42890000, 0x42890000, KB(64)),   /* ENET */               \
@@ -68,7 +80,7 @@ typedef enum IRQn {
 #define RTOS_COMMON_IRQCHIP0_BITMAP1 \
 	IRQnMAP(LPUART2_IRQn)
 #define RTOS_AUDIO_IRQCHIP0_BITMAP1 \
-	0
+	IRQnMAP(LPI2C1_IRQn)
 #define RTOS_AVB_IRQCHIP0_BITMAP1 \
 	RTOS_AUDIO_IRQCHIP0_BITMAP1
 #define RTOS_INDUS_IRQCHIP0_BITMAP1 \
@@ -80,7 +92,7 @@ typedef enum IRQn {
 #define RTOS_COMMON_IRQCHIP0_BITMAP2 \
 	IRQnMAP(TPM2_IRQn)
 #define RTOS_AUDIO_IRQCHIP0_BITMAP2 \
-	0
+	IRQnMAP(SAI1_IRQn)
 #define RTOS_AVB_IRQCHIP0_BITMAP2 \
 	RTOS_AUDIO_IRQCHIP0_BITMAP2
 #define RTOS_INDUS_IRQCHIP0_BITMAP2 \
@@ -130,7 +142,8 @@ typedef enum IRQn {
 #define RTOS_COMMON_IRQCHIP1_BITMAP2 \
 	IRQnMAP(SW_MB_R_IRQn)
 #define RTOS_AUDIO_IRQCHIP1_BITMAP2 \
-	0
+	IRQnMAP(SAI2_IRQn) | \
+	IRQnMAP(SAI3_IRQn)
 #define RTOS_AVB_IRQCHIP1_BITMAP2 \
 	RTOS_AUDIO_IRQCHIP1_BITMAP2 | \
 	IRQnMAP(ENET_QOS_PMT_IRQn) | \
diff --git a/configs/arm64/imx93-harpoon-zephyr-audio.c b/configs/arm64/imx93-harpoon-zephyr-audio.c
new file mode 100644
index 00000000..47ad8110
--- /dev/null
+++ b/configs/arm64/imx93-harpoon-zephyr-audio.c
@@ -0,0 +1,23 @@
+/*
+ * i.MX93 target - Zephyr audio config
+ *
+ * Copyright 2023 NXP
+ *
+ * Authors:
+ *  Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
+ *
+ * This work is licensed under the terms of the GNU GPL, version 2.  See
+ * the COPYING file in the top-level directory.
+ */
+
+#include "cell-helper.h"
+#include "imx93-harpoon-rtos.h"
+
+/* Name, cores */
+#define CONFIG_INMATE_NAME		"zephyr"
+#define CONFIG_INMATE_CORE_BITMAP	(0b10)
+
+/* Base, memory & peripherals, GIC, IVSHMEM PCI */
+#include "imx93-harpoon-config-audio.h"
+
+#include "cell-create.h"
-- 
2.34.1

